SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc

```diff
@@ -1,461 +1,673 @@
+
+  const std::string messageCategory_;
+
+  std::unordered_map<Index_t, float> m_detIdToTotalSimEnergy;  // keep track of cell normalizations
+  std::unordered_multimap<Barcode_t, Index_t> m_simHitBarcodeToIndex;
+
+  /** The maximum bunch crossing BEFORE the signal crossing to create
+      TrackinParticles for. Use positive values. If set to zero no
+      previous bunches are added and only in-time, signal and after bunches
+      (defined by maximumSubsequentBunchCrossing_) are used.
+  */
+  const unsigned int maximumPreviousBunchCrossing_;
+  /** The maximum bunch crossing AFTER the signal crossing to create
+      TrackinParticles for. E.g. if set to zero only
+      uses the signal and in time pileup (and previous bunches defined by the
+      maximumPreviousBunchCrossing_ parameter).
+  */
+  const unsigned int maximumSubsequentBunchCrossing_;
+
+  const edm::InputTag simTrackLabel_;
+  const edm::InputTag simVertexLabel_;
+  edm::Handle<std::vector<SimTrack> > hSimTracks;
+  edm::Handle<std::vector<SimVertex> > hSimVertices;
+
+  std::vector<edm::InputTag> collectionTags_;
+  edm::InputTag genParticleLabel_;
+  /// Needed to add HepMC::GenVertex to SimVertex
+  edm::InputTag hepMCproductLabel_;
+
+  const double minEnergy_, maxPseudoRapidity_;
+
+ public:
+  struct OutputCollections {
+    std::unique_ptr<SimClusterCollection> pSimClusters;
+    std::unique_ptr<CaloParticleCollection> pCaloParticles;
+  };
+
+  struct calo_particles {
+    std::vector<uint32_t> sc_start_;
+    std::vector<uint32_t> sc_stop_;
+
+    void swap(calo_particles& oth) {
+      sc_start_.swap(oth.sc_start_);
+      sc_stop_.swap(oth.sc_stop_);
+    }
+
+    void clear() {
+      sc_start_.clear();
+      sc_stop_.clear();
+    }
+  };
+
+ private:
+  const HGCalTopology* hgtopo_[2];
+  const HGCalDDDConstants* hgddd_[2];
+  const HcalDDDRecConstants* hcddd_;
+  OutputCollections output_;
+  calo_particles m_caloParticles;
+};
+
+/* Graph utility functions */
+
+namespace {
+template <typename Edge, typename Graph, typename Visitor>
+void accumulateSimHits_edge(Edge& e, const Graph& g, Visitor* v) {
+  auto const edge_property = get(edge_weight, g, e);
+  v->total_simHits += edge_property.simHits;
+  IfLogDebug(DEBUG, messageCategory_)
+      << " Examining edges " << e << " --> particle " << edge_property.simTrack->type() << "("
+      << edge_property.simTrack->trackId() << ")"
+      << " with SimClusters: " << edge_property.simHits
+      << " Accumulated SimClusters: " << v->total_simHits << std::endl;
+}
+template <typename Vertex, typename Graph>
+void print_vertex(Vertex& u, const Graph& g) {
+  auto const vertex_property = get(vertex_name, g, u);
+  IfLogDebug(DEBUG, messageCategory_) << " At " << u;
+  // The Mother of all vertices has **no** SimTrack associated.
+  if (vertex_property.simTrack)
+    IfLogDebug(DEBUG, messageCategory_) << " [" << vertex_property.simTrack->type() << "]"
```

Comment:
Maybe better to write the whole message into a stringstream and then print it all at once, avoiding breaking it up because of the conditional?